esx_service/utils/auth_data.py

```diff
@@ -0,0 +1,381 @@
+        self.conn.row_factory = sqlite3.Row
+
+        if not self.conn:
+            raise DbConnectionError(self.db_path)
+
+    def create_tables(self):
+        """
+        Create the 3 tables used for per-datastore authorization.  This function
+        should only be called once per datastore.  It will raise an exception if
+        the schema file isn't accessible or the tables already exist.
+        """
+        
+        with open('./docker-volume-auth-schema.sql') as f:
+            sql = f.read()
+            self.conn.executescript(sql)
+            self.conn.commit()
+
+    def create_tenant(self, name, description, default_datastore, default_privileges, 
+                      vms, privileges):
+        """
+        Create a tenant in the database. A tenant id will be auto-generated and
+        returned. vms are (vm_id, vm_name) pairs. Privileges are dictionaries
+        with keys matching the row names in the privileges table. Tenant id is
+        filled in for both the vm and privileges tables.
+        """
+
+        # Create the entry in the tenants table
+        tenant = DockerVolumeTenant(name, description, default_datastore,
+                                    default_privileges, vms, privileges)
+        id = tenant.id
+        self.conn.execute(
+            "insert into tenants(id, name, description, default_datastore) values (?, ?, ?, ?)",
+            (id, name, description, default_datastore)
+        )
+
+        # Create the entries in the vms table
+        vms = [(vm_id, vm_name, id) for (vm_id, vm_name) in vms]
+        #print vms
+        if vms:
+            self.conn.executemany(
+              "insert into vms(vm_id, vm_name, tenant_id) values (?, ?, ?)",
+              vms
+            )
+
+        # Create the entries in the privileges table
+        # TODO: data scrubbing, ensure all keys exist
+        for p in default_privileges:
+            p['tenant_id'] = id
+        self.conn.executemany(
+            """
+            insert into privileges values
+            (:tenant_id, :datastore, :global_visibility, :create_volume,
+             :delete_volume, :mount_volume, :max_volume_size, :usage_quota)
+            """,
+            default_privileges
+        )
+
+        for p in privileges:
+            p['tenant_id'] = id
+        self.conn.executemany(
+            """
+            insert into privileges values
+            (:tenant_id, :datastore, :global_visibility, :create_volume,
+             :delete_volume, :mount_volume, :max_volume_size, :usage_quota)
+            """,
+            privileges
+        )
+        self.conn.commit()
+        return tenant
+    
+    def list_tenants(self):
+        """
+            Return a list of DockerVolumeTenants objects.
+        """
+        tenant_list = []
+        cur = self.conn.execute(
+           "select * from tenants"
+        )
+        result = cur.fetchall()
+       
+        for r in result:
```

Comment:
yes, and it's usually a join. something like 'SELECT tenant, vm FROM tenants, vms WHERE tenants.id == vms.tenant_id`.
I realize DockerVolumeTenant constructor expect specific params, but it seems wrong to do work instead of delegating to SQLlite. Let's review offline and decide if it needs to change in this PR. 
Let's review this section offline.